src/Generating/PieceGeneratorBFSTree.cpp

```diff
@@ -0,0 +1,345 @@
+	{
+		a_OutConnectors.push_back(
+			cFreeConnector(res, StartingPiece->RotateMoveConnector(*itr, Rotation, a_BlockX, BlockY, a_BlockZ))
+		);
+	}
+
+	return cPlacedPiecePtr(res);
+}
+
+
+
+
+
+bool cPieceGeneratorBFSTree::TryPlacePieceAtConnector(
+	const cPlacedPiece & a_ParentPiece,
+	const cPiece::cConnector & a_Connector,
+	cPlacedPieces & a_OutPieces,
+	cPieceGeneratorBFSTree::cFreeConnectors & a_OutConnectors
+)
+{
+	// Get a list of available connections:
+	cConnections Connections;
+	int WantedConnectorType = -a_Connector.m_Type;
+	cPieces AvailablePieces = m_PiecePool.GetPiecesWithConnector(WantedConnectorType);
+	Connections.reserve(AvailablePieces.size());
+	Vector3i ConnPos = cPiece::cConnector::AddDirection(a_Connector.m_Pos, a_Connector.m_Direction);  // The position at which the new connector should be placed - 1 block away from the current connector
+	int WeightTotal = 0;
+	for (cPieces::iterator itrP = AvailablePieces.begin(), endP = AvailablePieces.end(); itrP != endP; ++itrP)
+	{
+		// Get the relative chance of this piece being generated in this path:
+		int Weight = m_PiecePool.GetPieceWeight(a_ParentPiece, a_Connector, **itrP);
+		if (Weight <= 0)
+		{
+			continue;
+		}
+
+		// Try fitting each of the piece's connector:
+		cPiece::cConnectors Connectors = (*itrP)->GetConnectors();
+		auto verticalLimit = (*itrP)->GetVerticalLimit();
+		for (cPiece::cConnectors::iterator itrC = Connectors.begin(), endC = Connectors.end(); itrC != endC; ++itrC)
+		{
+			if (itrC->m_Type != WantedConnectorType)
+			{
+				continue;
+			}
+			// This is a same-type connector, find out how to rotate to it:
+			int NumCCWRotations = cPiece::cConnector::GetNumCCWRotationsToFit(a_Connector.m_Direction, itrC->m_Direction);
+			if ((NumCCWRotations < 0) || !(*itrP)->CanRotateCCW(NumCCWRotations))
+			{
+				// Doesn't support this rotation
+				continue;
+			}
+
+			// Check if the piece's VerticalLimit allows this connection:
+			if ((verticalLimit != nullptr) && (!verticalLimit->CanBeAtHeight(ConnPos.x, ConnPos.z, ConnPos.y - itrC->m_Pos.y)))
+			{
+				continue;
+			}
+
+			if (!CheckConnection(a_Connector, ConnPos, **itrP, *itrC, NumCCWRotations, a_OutPieces))
+			{
+				// Doesn't fit in this rotation
+				continue;
+			}
+			// Fits, add it to list of possibile connections:
+			Connections.push_back(cConnection(**itrP, *itrC, NumCCWRotations, Weight));
+			WeightTotal += Weight;
+		}  // for itrC - Connectors[]
+	}  // for itrP - AvailablePieces[]
+	if (Connections.empty())
+	{
+		// No available connections, bail out
+		return false;
+	}
+	ASSERT(WeightTotal > 0);
+
+	// Choose a random connection from the list, based on the weights:
+	int rnd = (m_Noise.IntNoise3DInt(a_Connector.m_Pos.x, a_Connector.m_Pos.y, a_Connector.m_Pos.z) / 7) % WeightTotal;
+	size_t ChosenIndex = 0;
+	for (cConnections::const_iterator itr = Connections.begin(), end = Connections.end(); itr != end; ++itr, ++ChosenIndex)
```

Comment:
I had no idea something like that existed. And anyway it doesn't seem to be constructible in iterations (pushing the individual weights one at a time) and even if it were, it would have resulted in an extra allocation for the weights buffer.